src/lib/SaferNames/Name.hs

```diff
@@ -4,288 +4,245 @@
-  let (names, vals) = unzip $ LM.assocs m
-  (unsafeMakePlainBinderList names, vals)
-
-unsafeMakePlainBinderList :: [RawName] -> PlainBinderList n l
-unsafeMakePlainBinderList names = case names of
-  [] -> unsafeCoerceB Empty
-  (name:rest) -> Nest (UnsafeMakeBinder name) $ unsafeMakePlainBinderList rest
-
--- === specialized traversals ===
-
-type NameSet n = S.Set (Name n)
-
-fmapNames :: HasNamesE e => Scope b -> (Name a -> Name b) -> e a -> e b
-fmapNames scope f e = runIdentity $ traverseNamesE scope t e
-  where t = newNameTraversal (pure . f)
-
-foldMapNames :: (HasNamesE e, Monoid a) => (Name n -> a) -> e n -> a
-foldMapNames f e =
-  execWriter $ traverseNamesE unitScope t e
-  where t = newNameTraversal \name -> tell (f name) >> return unitName
-
-freeNames :: HasNamesE e => e n -> NameSet n
-freeNames e = foldMapNames S.singleton e
-
--- should be equivalent to `\ext -> fmapNames (injectName ext)`
-injectNamesL :: HasNamesE e => FreshExt n l -> e n -> e l
-injectNamesL _ = unsafeCoerceE
-
--- TODO: consider returning one of the `Name (n:=>:l)` names found in the
--- `Nothing` case, or even the full set.
-projectNamesL :: HasNamesE e => Scope n -> Scope (n:=>:l) -> e l -> Maybe (e n)
-projectNamesL scope scopeFragment e = traverseNamesE scope t e
-  where t = newNameTraversal \name -> case projectName scopeFragment name of
-                                        Left name' -> Just name'
-                                        Right _    -> Nothing
+scopeAsPlainScope :: Scope n -> PlainScope n
+scopeAsPlainScope (Scope m) = PlainScope $ nameMapNames m
 
 -- === common scoping patterns ===
 
 data Abs (binder::B) (body::E) (n::S) where
   Abs :: binder n l -> body l -> Abs binder body n
 
-data FreshAbs (binder::B) (body::E) (n::S) where
-  FreshAbs :: FreshExt n l -> binder n l -> body l -> FreshAbs binder body n
-
 data NestPair (b1::B) (b2::B) (n::S) (l::S) where
   NestPair :: b1 n l -> b2 l l' -> NestPair b1 b2 n l'
 
 data Nest (binder::B) (n::S) (l::S) where
   Nest  :: binder n h -> Nest binder h l -> Nest binder n l
   Empty ::                                  Nest binder n n
 
-type PlainBinderList = Nest PlainBinder
-
-infixl 7 :>
-data AnnBinderP (b::B) (ann ::E) (n::S) (l::S) =
-  (:>) (b n l) (ann n)
-  deriving (Show)
+type EmptyAbs b = Abs b UnitE :: E
+pattern EmptyAbs :: b n l -> EmptyAbs b n
+pattern EmptyAbs bs = Abs bs UnitE
+
+-- === Injections and projections ===
+
+newtype Ext (n::S) (l::S) = ExtVal { fromExtVal :: NameSet (n:=>:l) }
+
+instance Category Ext where
+  id = ExtVal emptyNameSetFrag
+  ext1 . ext2 = ExtVal $ concatNameSets (fromExtVal ext2) (fromExtVal ext1)
+
+class BindsNames (b :: B) where
+  toExtVal :: b n l -> Ext n l
+
+instance BindsNames Ext where
+  toExtVal = id
+
+toNameSet :: BindsNames b => b n l -> NameSet (n:=>:l)
+toNameSet b = fromExtVal $ toExtVal b
+
+inject :: BindsNames b => InjectableE e => Distinct l => b n l -> e n -> e l
+inject ext x = injectNames (toNameSet ext) x
```

Comment:
At some point we might want to decorate `injectNames` with `{-# INLINE #-}` to force GHC to inline the definition and make it realize that it doesn't actually have to compute the first argument anywhere.

Reply:
Laziness doesn't help use there?